### PR TITLE
Add failure reporting for async ckpt saver.

### PR DESCRIPTION
**Pull request by BalaBalaYi**
### What changes were proposed in this pull request?

Use 'master_client' to report ckpt failures in the async thread implementation of async ckpt saver.

### Why are the changes needed?

To catch and expose possible problems during saving ckpt in async thread.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

File: dlrover/python/elastic_agent/torch/training.py [hunk 1]
```diff
@@ -559,6 +552,9 @@ def _initialize_workers(self, worker_group):
                 # We need to register handler after starting workers because
                 # the PContext start_worker will overwrite the handler.
                 AsyncCheckpointSaver.register_signal_handler()
+
+                # need master client to report unexpected failures in saver
+                AsyncCheckpointSaver.register_signal_handler()
```
Comment (workingloong) [inline dlrover/python/elastic_agent/torch/training.py:557]: The line is repeated with L554.

**Comment by Eugene1518** [comment]
I have received your email.Thanks~

File: dlrover/python/elastic_agent/torch/training.py (hunk 1 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/training.py:557]: thx a lot

**Review by workingloong — APPROVED** [review]
LGTM

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 2]
```diff
@@ -522,20 +535,60 @@ def _sync_shm_to_storage(self):
...
+        try:
+            error = ProcessError(
+                self._node_rank,
+                -1,
+                error_full_msg,
+                str(datetime.fromtimestamp(int(time.time()))),
+            )
+
+            self._master_client.report_failures(
+                json.dumps(error.__dict__),
```
Comment (workingloong) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py:585]: The ProcessError can extend the JsonSerializable and we can use `error.to_json()`.

https://github.com/intelligent-machine-learning/dlrover/blob/8862fa56829bca1d003b425a802340a7840492dd/dlrover/python/common/serialize.py#L28-L35

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 3]
```diff
@@ -522,20 +535,60 @@ def _sync_shm_to_storage(self):
...
+        if not error_msg:
+            error_msg = "Unknown"
+        error_full_msg = "Async checkpointer saver got failure:" + error_msg
+
+        try:
+            error = ProcessError(
+                self._node_rank,
+                -1,
+                error_full_msg,
+                str(datetime.fromtimestamp(int(time.time()))),
```
Comment (workingloong) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py:581]: `datetime.now().strftime("%m/%d/%Y %H:%M:%S")` is more common.

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 4]
```diff
@@ -522,20 +535,60 @@ def _sync_shm_to_storage(self):
...
+        if not self._master_client:
+            logger.warning(
+                "Skip ckpt saver failure reporting for master "
+                "client hasn't setup."
+            )
+            return
+
+        if not error_msg:
+            error_msg = "Unknown"
+        error_full_msg = "Async checkpointer saver got failure:" + error_msg
```
Comment (workingloong) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py:574]: checkpointer  -> checkpoint

File: dlrover/python/elastic_agent/torch/ckpt_saver.py (hunk 2 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/ckpt_saver.py:585]: done

File: dlrover/python/elastic_agent/torch/ckpt_saver.py (hunk 3 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/ckpt_saver.py:581]: done

File: dlrover/python/elastic_agent/torch/ckpt_saver.py (hunk 4 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/ckpt_saver.py:574]: done

**Review by samplise — APPROVED** [review]

**Pull request merged**